portfolio/src/main/java/com/google/sps/data/TimeRangeGroupArrayList.java

```diff
@@ -0,0 +1,121 @@
+package com.google.sps.data;
+
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+
+/** Models an implementation of the TimeRangeGroup model using ArrayList. */
+public class TimeRangeGroupArrayList implements TimeRangeGroup {
+
+  public List<TimeRange> allTimeRanges;
+  private Comparator<TimeRange> comparator;
+  private boolean ascending;
+
+  /**
+   * Adds all the input time ranges to the list of all time ranges. Also sorts the list of all
+   * ranges in the constructor.
+   */
+  public TimeRangeGroupArrayList(
+      Collection<TimeRange> timeRanges, Comparator<TimeRange> comparator, boolean ascending) {
+    allTimeRanges = new ArrayList<TimeRange>();
+    timeRanges.forEach(
+        (range) -> {
+          addTimeRange(range);
+        });
+    this.comparator = comparator;
+    this.ascending = ascending;
+    sortTimeRanges(comparator, ascending);
+  }
+
+  /**
+   * Adds a new time range to the list. This new range must be disjoint from all existing ranges.
+   * TODO(hollyyuqizheng): Depending on how the algorithms are implemented, we might need to add the
+   * merging logic in this method. For example, if a time range is deleted but then later added back
+   * to the list for some reason, this list will potentially need to be merged.
+   */
+  public void addTimeRange(TimeRange timeRange) {
+    for (TimeRange currentRange : allTimeRanges) {
+      if (currentRange.overlaps(timeRange)) {
+        throw new IllegalArgumentException(
+            "New time range to add cannot overlap with any existing time ranges");
+      }
+    }
+    allTimeRanges.add(timeRange);
+  }
+
+  /** Returns the array list of all time ranges. */
+  public List<TimeRange> getAllTimeRanges() {
```

Comment:
I think one issue with getAllTimeRanges is that the caller can modify the result, which can affect the internal state of ArrayListTimeRangeGroup. I think this structure is similar to HashSet in the method signatures, and you can consider exposing an iterator like hash set, rather than getAllTimeRanges. That way, callers can use the iterator to iterate over all the time ranges.

Alternatively, you can update this method so that it returns an ImmutableList rather than List.

Reference:
https://docs.oracle.com/javase/7/docs/api/java/util/HashSet.html

Reply:
Updated by returning an ImmutableList.